src/devices/vdml_ext_adi.c

```diff
@@ -139,68 +159,97 @@ int32_t ext_adi_analog_calibrate(uint8_t smart_port, uint8_t adi_port) {
 	}
 	adi_data_s_t* const adi_data = &((adi_data_s_t*)(device->pad))[adi_port];
 	adi_data->analog_data.calib = (int32_t)((total + 16) >> 5);
-	return_port(smart_port - 1, (int32_t)((total + 256) >> 9));
+	rtn = (int32_t)((total + 256) >> 9);
+
+	done:
+	return_port(smart_port - 1, rtn);
 }
 
 int32_t ext_adi_analog_read(uint8_t smart_port, uint8_t adi_port) {
+	int32_t rtn;
 	transform_adi_port(adi_port);
 	claim_port_i(smart_port - 1, E_DEVICE_ADI);
-	validate_type(device, adi_port, E_ADI_ANALOG_IN);
-	int32_t rtn = vexDeviceAdiValueGet(device->device_info, adi_port);
+	if ((rtn = validate_type(device, adi_port, E_ADI_ANALOG_IN)))
+		goto done;
```

Comment:
I left off the != 0 for compactness. I can add it if people think it's helpful.

This goto cleanup pattern is fairly common, and is a decent way of not duplicating the resource freeing logic.
The other way I can think of doing this is have claim_port return a handle and always call into a helper function, where we can use that to detect what resources we have to free.

Reply:
While I'm personally fine with the goto pattern, for the sake of maintainability and to prevent confusion for future devs I'd say calling return_port() might be more concise. It's not that big of a deal but it's also not in line with how other portions of the kernel are written.